src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -62,8 +63,37 @@ struct async_connect {
   grpc_endpoint** ep;
   grpc_closure* closure;
   grpc_channel_args* channel_args;
+  int64_t connection_handle;
+  bool connect_cancelled;
 };
 
+namespace {
+gpr_once g_tcp_client_posix_init = GPR_ONCE_INIT;
+size_t g_pending_connection_shards = 0;
```

Comment:
How about something like this to minimize the number of globals needed:

```
struct ConnectionShard {
  grpc_core::Mutex mu;
  absl::flat_hash_map<int64_t, async_connect*> pending_connections ABSL_GUARDED_BY(&mu);
};
std::vector<ConnectionShard>* g_connection_shards = nullptr;
```

This can also be initialized more simply:

```
size_t num_shards = std::max(2 * gpr_cpu_num_cores(), 1u);
g_connection_shards = new std::vector<ConnectionShard>(num_shards);
```

Reply:
Updated to use this struct.